packages/core/nano/src/types.ts

```diff
@@ -1,22 +1,139 @@
-export interface Consumer<T> {
-    notify: <K extends keyof T>(property: K, value: T[K]) => void;
-    notifyMany: <K extends keyof T>(values: Pick<T, K>) => void;
+export interface Pending<T> {
+  kind: "Pending";
+  estimate?: T;
 }
 
-export interface Pending<T> {
-    kind: "Pending";
-    estimate?: T;
+/**
+ * The interface for an object whose data can be bound to. We use this contract for
+ * components that want to expose their data and its changes to other components.
+ * An example of this would be Formula binding to a Tablero. In this scenario,
+ * Tablero is expected to implement IProducer and expose some of its data, given
+ * a reference, through getDataResult. Formula is expected to implement IConsumer
+ * and register itself as an observer to Tablero.
+ *
+ * Any component that implements IProducer is expected to provide some registration
+ * functionality and to notify observers whenever the data they are bound to changes.
+ */
+export interface IProducer<T> {
+  /**
+   * Binds the given observer to the component associated with this Producer.
+   * @param consumer - The object to start listening to any changes in the bound data.
+   */
+  addConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Unbinds the given observer from the component associated with this Producer.
+   * @param consumer - The observer to unregister from the Producer.
+   */
+  removeConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Return IDataResult for the given dataBindingReference, containing the data value
+   * and some additional metadata.
+   * @param dataBindingReference - An instance of dataBindingReference, a reference
+   * to some data in the bound component.
+   */
+  read<K extends keyof T>(consumer: IConsumer<T>, property: K, ...args: any[]): T[K] | Pending<T[K]>;
+
+  /**
+   * Invoked when this data provider is deleted.
+   */
+  onDeleted(): void;
+}
+
+/**
+ * The interface for an object that can bind to an IProducer.
+ * One example of this provider-observer relationship is a Formula binding
+ * to Tablero. In this scenario, Formula is an IConsumer listening to changes
+ * in the IProducer, Tablero.
+ *
+ * Any object that implements IConsumer is expected to provide a
+ * callback whenever the component it is bound to changes in value and a reference
+ * to the data that the observer is bound to.
+ */
+export interface IConsumer<T> {
+  /**
+   * Invoked whenever the data this object is bound to is changed.
+   */
+  notify: <K extends keyof T>(property: K, value: T[K]) => void;
+
+  /**
+   * Invoked whenever the provider this is bound to is deleted.
+   */
+  onProducerDeleted: () => void;
```

Comment:
What is the purpose of the onProducerDeleted callback?  In other systems I've worked on, I had a similar notification that told a consumer when a producer became fixed (i.e., wouldn't change again), giving the consumer an opportunity to drop any tracking related state.

Reply:
How are consumers notified when a producer is deleted? In a more practical example, say we have a Formula bound to the name of a table. When the table is deleted, we don't want the formula to show the table name anymore, we probably want it to show some broken reference UI. Also, we'd want to make sure the producer is unregistered from the DataBindingService (not a service we're adding into this repo). Can those concepts still be addressed without onDeleted?

Reply:
I think the implicit assumption here is that all producers that could be deleted are actually some property nested in another producer, and that a deletion is just a notification from the outer producer. Something like:
```ts
type FormulaRootContext = IProducer< { TableName: IProducer<ITable> | MissingRef } >;
```
The formula would be subscribed to the root context producer and when table `TableName` is deleted it gets notified that the property has changed its value from a table to 'missing', for some representation of missing.

Reply:
One idea is to use the revoker pattern:
```ts
class RevokableProducer<T> implements IProducer<T>, IConsumer<T> {
    private consumers = new Set<IConsumer<T>>();

    constructor(private readonly producer: IProducer<T>) {}

    private get isRevoked() { return !this.consumers; }

    public read(consumer, ...) {
        if (this.isRevoked) { return new Error(`Not available.`); }
        this.consumers.add(consumer);

        return this.producer.read(this, ..);
    }

    public removeConsumer(consumer) {
        this.consumers.delete(consumer);
        if (this.consumers.count === 0) { this.producer.unsubscribe(this); }
    }

    public notify(...) {
        assert(!this.revoked);
        for (const consumer of consumers) { consumer.notify(...); }
    }

    public revoke() {
        this.revoked = true;
        this.notify(this, '*', new Error('Not available.'));      // Invalidate all properties from this producer
        this.consumers = undefined;
        this.producer.unsubscribe(this);
    }   
}
```

Reply:
I see, that makes sense. So we just treat deletion as another property change, albeit a catastrophic one?

Reply:
Right.  You can also probably get more mileage out of the wrapper object (e.g., a general adapter for IProducers that knows how to create IDataResults for your app's internal use.)